### PR TITLE
bauhaus combobox : ensure full vertical align for real

### PR DESCRIPTION
**Pull request by AlicVB**
this is a follow up of @TurboGit work on #11351 : 
this ensure vertical alignment of text in *all* situation, whatever the widget height (well, if I've done it right :))

I've done that by computing the height of the text to be sure to align it correctly.

Here's how it looks :
Capture d’écran_2022-03-17_19-16-13
Line 1 : master before #11351 
Line 2 : master with #11351 
Line 3 : master with this PR
Line 4 : same but with a "forced" height to something big (just as a proof that it works)

Note that the text is aligned the same way as labels

There's still a small issue thought : currently bauhaus combobox apply an hardcoded margin of 1 px to the top only, so for small height with visible background-color, one may say that we should move the text 1 px down... but here the problem is clearly the hardcoded margin that should not be here and certainly not applied to the top only !
So next step is to remove that margin and to implement css margins instead... but not for this evening ;)

### CODE REVIEW

**Review by TurboGit — CHANGES_REQUESTED** [review]
Nice of the GUI part, alignment is good now.

There is one regression though, the hover detection is slightly off. That is, click to open a combo and fly over the items you'll see that it highlight the items a bit lower than the mouse position. You've probably missed the adjustment for the highlight computation.

**Comment by TurboGit** [comment]
As a hint, on my previous attempt the issue was in line 1799:

```
      const int active = darktable.bauhaus->end_mouse_y >= 0
                 ? ((darktable.bauhaus->end_mouse_y - INNER_PADDING) / (darktable.bauhaus->line_height))
                 : d->active;
```

File: src/bauhaus/bauhaus.c [hunk 1]
```diff
@@ -1648,15 +1653,15 @@ static void dt_bauhaus_draw_quad(dt_bauhaus_widget_t *w, cairo_t *cr)
     switch(w->type)
     {
       case DT_BAUHAUS_COMBOBOX:
-        cairo_translate(cr, width - darktable.bauhaus->quad_width * .5f, height * .4f);
```
Comment (Nilvus) [inline src/bauhaus/bauhaus.c:1651]: I had put that setting to have a smaller arrow, and nearly same size as classic combo (check on collapsible section of import module). The arrow is now a little too big.
But I've tested by putting back 0.4 and indeed, now it's not good as it break vertical align you succeed to implement.

It should be possible anyway to reduce a little the arrow. Could you update that?

**Review by Nilvus — CHANGES_REQUESTED** [review]

**Comment by AlicVB** [comment]
> Nice of the GUI part, alignment is good now.
> 
> There is one regression though, the hover detection is slightly off. That is, click to open a combo and fly over the items you'll see that it highlight the items a bit lower than the mouse position. You've probably missed the adjustment for the highlight computation.

hmmm... in my tests, the hover change is done exactly at the same position with this PR, master or master before your commit...
So I would say that it's not a regression but another thing to fix :)

**Comment by AlicVB** [comment]
bauhaus widgets now use css margin and padding properties.
For bauhaus widgets inside modules, I've set default top margin to 1px to mimic actual hardcoded UI.

**Comment by TurboGit** [comment]
> hmmm... in my tests, the hover change is done exactly at the same position with this PR, master or master before your commit...
> So I would say that it's not a regression but another thing to fix :)

Indeed, I see the same issue on master. I was sure I had fixed this. My testing is to start on the middle of a combo entry and check that the item below or above are highlighted after moving the same distance up or down.

**Comment by TurboGit** [comment]
> As a hint, on my previous attempt the issue was in line 1799:
> 
> ```
>       const int active = darktable.bauhaus->end_mouse_y >= 0
>                  ? ((darktable.bauhaus->end_mouse_y - INNER_PADDING) / (darktable.bauhaus->line_height))
>                  : d->active;
> ```

This is for the selected item, for the highlighted one (line 2023): 

      const int hovered = (darktable.bauhaus->mouse_y - INNER_PADDING) / darktable.bauhaus->line_height;

**Comment by AlicVB** [comment]
I've just seen some problem about mouse hovering and positions (in sliders esp.) so I put that PR to draft until I manage to fix everything !

**Comment by AlicVB** [comment]
OK, so in my tests, everything works now. Lot of hardcoded values / sizes have been removed.
- widgets take account of css padding + margin
- popup take account of padding only
- combobox popup item hovering is fixed
- a bug in slider popup, when moving the line far away from the origin point, is fixed too
- css has been adjusted to mimic previous behavior

This need more tests now, because it's very easy to break things here :) (hopefully it will be easier to fix issues now that the code is a little bit cleaner...)

File: src/bauhaus/bauhaus.c (hunk 1 above)
Comment (TurboGit) [reply src/bauhaus/bauhaus.c:1651]: Not that bad to me, it makes the collapsible section a bit more visible.

**Review by TurboGit — CHANGES_REQUESTED** [review]
The hovered widget is still not computed correctly, add this CSS:

```
.bauhaus_combo_popup
{
   padding-top: 20px;
   margin-top: 20px;
}
```

Open the filter lighttable combo for example and you'll see that the highlighted item is not the one under the mouse pointer.

File: src/bauhaus/bauhaus.c (hunk 1 above)
Comment (Nilvus) [reply src/bauhaus/bauhaus.c:1651]: Alic answer to me in Matrix and to change size correctly, that was not there. So, this change is good so.

**Comment by TurboGit** [comment]
I think the hovered computation should be:

`      const int hovered = (darktable.bauhaus->mouse_y - w->padding->top) / darktable.bauhaus->line_height;`

Looking at the `w->padding->top` of the widget I get 2. So it seems that this is not PX but mouse_y and line_height are in PX, so even with that we are still off.

**Comment by TurboGit** [comment]
I'll test your change, I was about to propose that:

```
diff --git a/src/bauhaus/bauhaus.c b/src/bauhaus/bauhaus.c
index 23314c642..cb78d07d8 100644
--- a/src/bauhaus/bauhaus.c
+++ b/src/bauhaus/bauhaus.c
@@ -1802,6 +1802,7 @@ static void dt_bauhaus_widget_accept(dt_bauhaus_widget_t *w)
   const int width = allocation_popup_window.width - w->padding->left - w->padding->right;
   const int height = allocation_popup_window.height - w->padding->top - w->padding->bottom;
   const int base_height = darktable.bauhaus->line_height + INNER_PADDING * 2.0f;
+  const GtkBorder *padding = darktable.bauhaus->popup_padding;
 
   switch(w->type)
   {
@@ -1810,7 +1811,7 @@ static void dt_bauhaus_widget_accept(dt_bauhaus_widget_t *w)
       // only set to what's in the filtered list.
       dt_bauhaus_combobox_data_t *d = &w->data.combobox;
       const int active = darktable.bauhaus->end_mouse_y >= 0
-                 ? ((darktable.bauhaus->end_mouse_y - INNER_PADDING) / (darktable.bauhaus->line_height))
+                 ? ((darktable.bauhaus->end_mouse_y - padding->top) / darktable.bauhaus->line_height)
                  : d->active;
 
       int k = 0, i = 0, kk = 0, match = 1;
@@ -1885,7 +1886,7 @@ static gboolean dt_bauhaus_popup_draw(GtkWidget *widget, cairo_t *crf, gpointer
   // dimensions of the popup
   GtkAllocation allocation;
   gtk_widget_get_allocation(widget, &allocation);
-  GtkBorder *padding = darktable.bauhaus->popup_padding;
+  const GtkBorder *padding = darktable.bauhaus->popup_padding;
   const int w2 = allocation.width - padding->left - padding->right;
   const int h2 = allocation.height - padding->top - padding->bottom;
 
@@ -2001,7 +2002,8 @@ static gboolean dt_bauhaus_popup_draw(GtkWidget *widget, cairo_t *crf, gpointer
       gboolean first_label = TRUE;
       gboolean show_box_label = TRUE;
       int k = 0, i = 0;
-      const int hovered = darktable.bauhaus->mouse_y / darktable.bauhaus->line_height;
+      const int hovered = (darktable.bauhaus->mouse_y - padding->top) / darktable.bauhaus->line_height;
+
       gchar *keys = g_utf8_casefold(darktable.bauhaus->keys, -1);
       const PangoEllipsizeMode ellipsis = d->entries_ellipsis;
       ht = darktable.bauhaus->line_height;
```

**Comment by AlicVB** [comment]
@TurboGit : I should have fixed it. For the hovering part, it was a stupid error as I have taken the padding of the widget instead of the one of the popup... For information, `darktable.bauhaus->mouse_y` should already take in account the padding (same for `end_mouse_y`
I've also fixed an even more problematic issue with the click to select which wasn't selecting the right item... That was due to the fact that the selection routine was called 2 times, on button press and release, with different values...

**Review by TurboGit — APPROVED** [review]
All good to me now, thanks! That's a really nice achievement and it makes dt even more CSS configurable. Congrats!

**Comment by dterrahe** [comment]
This crashes hard when you assign a shortcut to a bauhaus, then hide the side panels, reenter darkroom and press the shortcut.

The reason for this is probably that the handling of this situation (which is quite tricky) got ignored. In this case the (hidden) widget is not instantiated or sized, so you can't just simply use the size of the widget itself for the popup.

**Comment by AlicVB** [comment]
hmm... indeed, I've not test this. I'll try to reproduce and see what I can do.
sorry to cause so much troubles !

**Pull request merged**